### PR TITLE
fix #30 Fit vertically/adjust to terminal height

### LINKED ISSUE #30: Feature: option to fit vertically instead of horizontally
Hello and thanks for this very nice program!

It seems like big images are resized to fit the width of the console, but I would prefer to see the whole image without scrolling.  Could you add a way to control this?

Alternatively, an option (or changing the default behaviour) to fit horizontally or vertically depending on the picture (i.e. use the most screen without overflowing) would also work for me.

### PR DESCRIPTION
**Pull request by boretom**
**What kind of change does this PR introduce?** (check at least one)

- [ ] Bugfix
- [x] Feature
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

If yes, please describe the impact and migration path for existing applications:

**The PR fulfills these requirements:**

- [x] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix #xxx[,#xxx]`, where "xxx" is the issue number)
- [ ] All tests are passing
- [ ] New/updated tests are included

If adding a **new feature**, the PR's description includes:
- [x] This PR allows to adjust the output to terminal height instead of terminal width. The default behaviour stays the same only if you pass `-H` it changes. Passing `-H` or `-H 0`  will adjust the image to the terminal height, passing a number is adjust to that height.

### CODE REVIEW

**Comment by boretom** [comment]
Just as I submitted the PR I realized that I haven't considered both `-w` and `-H` passed. I only ever used one or the other. So you may want to hold on looking at that PR.

**Comment by posva** [comment]
Thank you for this!
Yeah, I think having both -w and -H could yield an error

**Comment by boretom** [comment]
Morning @posva, what behaviour would you prefer if both `-w` and `-H` are passed? I see the following options

1. only use whatever parameter is passed first, ignore the second
2. only use whatever parameter is passed last, ignore the one before
3. abort with error message, "only either -w or -H parameter is allowed"
4. allow both at the same time and resize width and height

Personally I would prefer number 3

**Comment by posva** [comment]
Aborting with an error message sounds good :+1:

**Comment by boretom** [comment]
Hi, I added the check and abort if both are passed.

Additionally I did reorganise the man page file a bit so the options are sorted alphabetically. Plus reorganised the synopsis a bit. If you don't like it I'll of course restore the old man page + add the `-H` option.

Related to the man page I got some questions:

- are `-t` and `-c` arguments which can be used together?
- in `-w` it's written: "[...] by default catimg will use the terminal width". What does that mean?
- would it make sense to add an `EXAMPLE` section? To clarify that, eg. `-w` without a positiv integer is not allowed - although no error is printed right now. So even `-w -r 1` works kind of but not really. In that case `-r 1` is not working because `-r` is assigned as value to `-w`. Because `-w` expects an argument.

File: man/catimg.1 [hunk 1]
```diff
@@ -1,41 +1,43 @@
...
+[\fB-hct\fP] [\fB-w width\fP | \fB-H height\fP] [\fB-l loops\fP] [\fB-r resolution\fP] image
 
 .SH DESCRIPTION
 .B catimg
 is a little program written in C with no dependencies that prints images in the terminal. It supports JPEG, PNG and GIF formats.
 
 .SH OPTIONS
 .TP
+\fB\-c\fR
+Convert colors to a restricted palette. This is useful when terminal only support a limited set of colors. This transformation should be more accurate than the one performed by the terminal.
```
Comment (posva) [inline man/catimg.1:17]: I'm using this occasion to add a few changes to my own code 😄 

```suggestion
Coerce colors to a restricted palette. Allows terminals with limited color support to render a more accurate version of the image than if it was directly done by the terminal itself.
```

File: man/catimg.1 [hunk 2]
```diff
@@ -1,41 +1,43 @@
...
 Possible values 1 or 2. Force the resolution of the image. By default catimg will check if rendering in higher resolution is possible and do so or use the lower one.
 .TP
-\fB\-c\fR
-Convert colors to a restricted palette. This is useful when terminal only support a limited set of colors. This transformation should be more accurate than the one performed by the terminal.
-.TP
 \fB\-t\fR
 Disables true color (24-bit) support, falling back to 256 color escape codes instead.
+.TP
+\fB\-w\fR WIDTH
+Specify the width of the displayed image, by default catimg will use the terminal width.
```
Comment (posva) [inline man/catimg.1:35]:
```suggestion
Specify the width of the displayed image. When not provided, catimg will use the terminal width.
```

File: man/catimg.1 [hunk 3]
```diff
@@ -1,41 +1,43 @@
...
 
 .SH BUGS
 Please report any bugs to https://github.com/posva/catimg/issues.
 
 .SH AUTHORS
 catimg was written by Eduardo San Martin Morote (https://github.com/posva)
 .LP
-This manual page was written by Jonathan Carter <jcarter@linux.com> and Eduardo San Martin Morote
+This manual page was written by Jonathan Carter <jcarter@linux.com> and Eduardo San Martin Morote and extended by Thomas Kupper
```
Comment (posva) [inline man/catimg.1:43]:
```suggestion
This manual page was written by Jonathan Carter <jcarter@linux.com>, Eduardo San Martin Morote and Thomas Kupper
```

File: src/catimg.c [hunk 4]
```diff
@@ -67,14 +71,39 @@ int main(int argc, char *argv[])
     int c;
     opterr = 0;
 
-    uint32_t cols = 0, precision = 0;
+    uint32_t cols = 0, rows = 0, precision = 0;
+    uint32_t maxCols = 0, maxRows = 0;
```
Comment (posva) [inline src/catimg.c:75]: let's use snake case since everything is in snake case

Comment (boretom) [reply src/catimg.c:75]: *after looking up what that is* sure I'll correct that with the other suggested changes

**Comment by boretom** [comment]
I did test it on macOS 10.15.5, macOS 10.14.6, Linux Mint 19.3, Linux Mint 20, Raspberry Pi 3B+ with Debian Buster and FreeBSD 12.1.

**Comment by boretom** [comment]
Oh, and no hurry at all. I'm old enough to know that the priorities can vary greatly.

**Comment by posva** [comment]
Oh wow, you already tested it in so many devices. After the needed changes, it should be good for a merge!

**Comment by boretom** [comment]
Testing is a big word :) ... it compiles and works fine for me on the mentioned platforms/distros with different images and two animated gifs.

I haven't done any testing on Windows. It would be interesting because according to [StackOverflow](https://stackoverflow.com/questions/6812224/getting-terminal-size-in-c-for-windows) getting `Y` the way I did it (same as for `X`) gives the buffer height but not the terminal window height.

Btw: It also compiles on Haiku OS R1/Beta2 but the output in the ~~BeOS~~ Haiku OS terminal doesn't look right. But if I connect to the Haiku box using SSH from my macOS it looks correct. So it's the Haiku OS terminal which behaves badly. I'll drop a note to one of their mailing list but not this week.

Btw2: It compiles and works on Fedora 32 (all the OS except macOS & FreeBSD run on a Intel NUC) but you would have to move two variable definitions (`color_map` and `yuv_color_map`) out of `sh_color.h` to `sh_color.c`. Otherwise the linker complains about them being defined twice (I read one should define variables in header files for this reason). One of them - `yuv_color_map` - is defined again in `sh_color.c` so the one in `sh_color.h` can be deleted.
If you like I can make that change too.

File: src/catimg.c [hunk 5]
```diff
@@ -67,14 +71,39 @@ int main(int argc, char *argv[])
...
     opterr = 0;
 
-    uint32_t cols = 0, precision = 0;
+    uint32_t cols = 0, rows = 0, precision = 0;
+    uint32_t max_cols = 0, max_rows = 0;
     uint8_t convert = 0;
     uint8_t true_color = 1;
-    while ((c = getopt (argc, argv, "w:l:r:hct")) != -1)
+    uint8_t adjust_to_height = 0, adjust_to_width = 0;
+    float sc = 0, sr = 0;
```
Comment (posva) [inline src/catimg.c:79]: can we also use full names for these variables?

File: src/catimg.c [hunk 6]
```diff
@@ -111,18 +140,31 @@ int main(int argc, char *argv[])
...
 
     if (strcmp(file, "-") == 0) {
         img_load_from_stdin(&img);
     } else {
         img_load_from_file(&img, file);
     }
-    if (cols < img.width) {
-        float sc = cols/(float)img.width;
+    if (cols == 0 && rows == 0) {
+        sc = max_cols/(float)img.width;
```
Comment (posva) [inline src/catimg.c:153]: can you put a span around mathematical operators like in the rest of the codebase?

File: src/catimg.c (hunk 5 above)
Comment (boretom) [reply src/catimg.c:79]: that can be done

File: src/catimg.c (hunk 6 above)
Comment (boretom) [reply src/catimg.c:153]: What does "a span" mean? Like spaces before and after '/'?

**Comment by boretom** [comment]
I made the changes as I mentioned in the comments + what I guessed does "put span around math operators" mean.

**Comment by posva** [comment]
I realized there is a floating point exception when providing a `-H` option that is too big. It should not exceed original height like it does with original width. I won't be able to take a look for some time, so please feel free to take a look at it and submit a PR!

**Comment by boretom** [comment]
I see, I'll check on it and great a PR. The same does not happen with `-w`?
What number did you use to test (I tested with -1 and 100000) and on what OS?

**Comment by boretom** [comment]
And regarding upgrading homebrew. a) no hurry and b) `brew edit catimg` shows you the recipe for that package.

**Comment by boretom** [comment]
I pulled the merged `master` and also get an floating point exception. But not with the `feature/add-height` branch, strange. I'll investigate it this week.

**Pull request merged**